### PR TITLE
Connection type update

### PR DESCRIPTION
**Pull request by krwoodal**
So looking at Apollo looks like when they get `connection_ack` they re send the subscription. 

```
 case .connectionAck:
        acked = true
        writeQueue()
```

So did this so we can match enums rather then strings

### CODE REVIEW

**Review by rex-remind101 — APPROVED** [review]
guessing we'll need to add that write as well in a separate pr?

File: AutoGraph/SubscriptionRequest.swift [hunk 1]
```diff
@@ -74,6 +74,7 @@ public enum GraphQLWSProtocol: String {
     case data = "data"                                 // Server -> Client
     case error = "error"                               // Server -> Client
     case complete = "complete"                         // Server -> Client
+    case unknown
```
Comment (rex-remind101) [inline AutoGraph/SubscriptionRequest.swift:77]: might be an improvement if it included `unknown_received_payload(String)` and we put what we got in into String

Comment (rex-remind101) [reply AutoGraph/SubscriptionRequest.swift:77]: naming as well, so it's obvious this is a Server -> Client thing

Comment (rex-remind101) [reply AutoGraph/SubscriptionRequest.swift:77]: wups, in camelCase though `unknownReceivedPayload(String)`

Comment (rex-remind101) [reply AutoGraph/SubscriptionRequest.swift:77]: or `unknownResponse(String)`

Comment (krwoodal) [reply AutoGraph/SubscriptionRequest.swift:77]: So I don't believe you can have associated value with raw enums. I can change the name though

File: AutoGraph/AutoGraph.swift [hunk 2]
```diff
@@ -101,23 +101,23 @@ open class AutoGraph {
     /// immediate failure, in which case, please check the error received in the completion block.
     ///
     /// It will form a websocket connection if one doesn't exist already.
-    open func subscribe<R: Request>(_ request: R, completion: @escaping RequestCompletion<R.SerializedObject>) -> Subscriber? {
+    open func subscribe<R: Request>(_ request: R, connectionState: @escaping SubscriptionResponseHandler.WebSocketConnectionBlock, completion: @escaping RequestCompletion<R.SerializedObject>) -> Subscriber? {
```
Comment (krwoodal) [inline AutoGraph/AutoGraph.swift:104]: @rex-remind101 So thinking of a way to update the UI of a connection state change to let the user know if the web socket is reconnecting. I thought adding another block would work

File: AutoGraph/WebSocketClient.swift [hunk 3]
```diff
@@ -7,6 +7,7 @@ public typealias WebSocketConnected = (Result<Void, Error>) -> Void
 public protocol WebSocketClientDelegate: class {
     func didReceive(event: WebSocketEvent)
     func didReceive(error: Error)
+    func didChangeConnection(state: WebSocketClient.State)
```
Comment (krwoodal) [inline AutoGraph/WebSocketClient.swift:10]: @rex-remind101 way to monitor state change of the webSocket

File: AutoGraph/WebSocketClient.swift [hunk 4]
```diff
@@ -187,6 +193,7 @@ open class WebSocketClient {
             self.attemptReconnectCount -= 1
         }
         self.reconnecting = true
+        self.state = .reconnecting
```
Comment (rex-remind101) [inline AutoGraph/WebSocketClient.swift:196]: why was this added? should we remove `self.reconnecting = true` then?

Comment (krwoodal) [reply AutoGraph/WebSocketClient.swift:196]: So we can present to the user when we are reconnecting

**Review by rex-remind101 — COMMENTED** [review]
👍

**Review by rex-remind101 — APPROVED** [review]
